src/main/scala/com/fulcrumgenomics/umi/FilterConsensusReads.scala

```diff
@@ -0,0 +1,217 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.bam.Bams
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.umi.VanillaUmiConsensusCallerOptions._
+import com.fulcrumgenomics.util.{Io, ProgressLogger}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.reference.ReferenceSequenceFileWalker
+import htsjdk.samtools.util.SequenceUtil
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+
+@clp(
+  group = ClpGroups.Umi, description =
+    """
+      |Filters a BAM file of consensus reads generated by CallMolecularConsensusReads. Two kinds of
+      |filtering are performed:
+      |  1. Masking/filtering of individual bases in reads
+      |  2. Filtering out of reads
+      |
+      |Base-level filtering of masking is only applied if per-base tags are present. Read-level
+      |filtering is always applied.  When filtering reads, secondary alignments and supplementary
+      |records may be removed independently if they fail one or more filters; if either R1 or R2
+      |primary alignments fail a filter then all records for the template will be filtered out.
+      |
+      |The filters applied are as follows:
+      |  1. Bases with quality scores below min-base-quality are masked to Ns
+      |  2. Bases with fewer than min-reads contributing raw reads are masked to Ns
+      |  3. Bases with a consensus error rate (defined as the fraction of contributing reads that
+      |     voted for a different base than the consensus call) higher than max-base-error-rate
+      |     are masked to Ns
+      |  4. Reads with a proportion of Ns higher than max-no-call-fraction *after* per-base
+      |     filtering are filtered out.
+      |  5. Reads with fewer than min-reads contributing reads are filtered out
+      |  6. Reads with an average consensus error rate higher than max-read-error-rate are
+      |     filtered out
+      |
+      |In order to correctly filter reads in or out by template, if the input BAM is not queryname
+      |sorted or grouped it will be sorted into queryname order.  Likewise, in order to emit
+      |records with corrected NM, UQ and MD tags the reads are re-sorted into coordinate order
+      |for output.
```

Comment:
Output sort order is always `coordinate`, I'll make that clearer but would prefer not to make that an option for now.